doc/source/reference/c-api/array.rst

```diff
@@ -1555,60 +1555,60 @@ For all of these macros *arr* must be an instance of a (subclass of)
     :c:data:`NPY_ARRAY_OWNDATA`, :c:data:`NPY_ARRAY_ALIGNED`,
     :c:data:`NPY_ARRAY_WRITEABLE`, :c:data:`NPY_ARRAY_WRITEBACKIFCOPY`.
 
-.. c:function:: int PyArray_IS_C_CONTIGUOUS(PyObject *arr)
+.. c:function:: int PyArray_IS_C_CONTIGUOUS(PyArrayObject *arr)
 
     Evaluates true if *arr* is C-style contiguous.
 
-.. c:function:: int PyArray_IS_F_CONTIGUOUS(PyObject *arr)
+.. c:function:: int PyArray_IS_F_CONTIGUOUS(PyArrayObject *arr)
 
     Evaluates true if *arr* is Fortran-style contiguous.
 
-.. c:function:: int PyArray_ISFORTRAN(PyObject *arr)
+.. c:function:: int PyArray_ISFORTRAN(PyArrayObject *arr)
 
     Evaluates true if *arr* is Fortran-style contiguous and *not*
     C-style contiguous. :c:func:`PyArray_IS_F_CONTIGUOUS`
     is the correct way to test for Fortran-style contiguity.
 
-.. c:function:: int PyArray_ISWRITEABLE(PyObject *arr)
+.. c:function:: int PyArray_ISWRITEABLE(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* can be written to
 
-.. c:function:: int PyArray_ISALIGNED(PyObject *arr)
+.. c:function:: int PyArray_ISALIGNED(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is properly aligned on
     the machine.
 
-.. c:function:: int PyArray_ISBEHAVED(PyObject *arr)
+.. c:function:: int PyArray_ISBEHAVED(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is aligned and writeable
     and in machine byte-order according to its descriptor.
 
-.. c:function:: int PyArray_ISBEHAVED_RO(PyObject *arr)
+.. c:function:: int PyArray_ISBEHAVED_RO(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is aligned and in machine
     byte-order.
 
-.. c:function:: int PyArray_ISCARRAY(PyObject *arr)
+.. c:function:: int PyArray_ISCARRAY(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is C-style contiguous,
     and :c:func:`PyArray_ISBEHAVED` (*arr*) is true.
 
-.. c:function:: int PyArray_ISFARRAY(PyObject *arr)
+.. c:function:: int PyArray_ISFARRAY(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is Fortran-style
     contiguous and :c:func:`PyArray_ISBEHAVED` (*arr*) is true.
 
-.. c:function:: int PyArray_ISCARRAY_RO(PyObject *arr)
+.. c:function:: int PyArray_ISCARRAY_RO(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is C-style contiguous,
     aligned, and in machine byte-order.
 
-.. c:function:: int PyArray_ISFARRAY_RO(PyObject *arr)
+.. c:function:: int PyArray_ISFARRAY_RO(PyArrayObject *arr)
 
     Evaluates true if the data area of *arr* is Fortran-style
     contiguous, aligned, and in machine byte-order **.**
 
-.. c:function:: int PyArray_ISONESEGMENT(PyObject *arr)
+.. c:function:: int PyArray_ISONESEGMENT(PyArrayObject *arr)
```

Comment:
These macros are all defined in terms of `PyArray_CHKFLAGS`, which takes a `const PyArrayObject *`, so I think all these need to take a `const PyArrayObject *` as well. I think whether or not someone uses const in their code only matters in C++ though so for C code what you have here is fine.

https://github.com/numpy/numpy/blob/dd3e7d93b836e965cba97be8b0a31daaa2dfeaeb/numpy/_core/include/numpy/ndarraytypes.h#L1572-L1576

Reply:
:+1: especially if we add it in the first part, would be nice to also do it here.

We added the `const` to a bunch of arguments, IIRC, so there are likely more missing (but no need to worry here).
For arguments like this, it is not a problem to make them const. It allows C++ users to use const, but doesn't require them (or change anything).
Problems do come with things arrays like `const char**` and/or `char *const *`, though.
Also Cython is strict about it (so the `__init__.pyd` doesn't declare it const.)